utility/publish.go

```diff
@@ -0,0 +1,316 @@
+	}
+	log.Infof("Uploaded apk version: %d", apk.VersionCode)
+	return apk, nil
+}
+
+// updates the listing info of a given release.
+func updateListing(configs config.Configs, release *androidpublisher.TrackRelease) error {
+	log.Debugf("Checking if updating listing is required, whats new dir is '%v'", configs.WhatsnewsDir)
+	if configs.WhatsnewsDir != "" {
+		fmt.Println()
+		log.Infof("Update listing started")
+
+		recentChangesMap, err := readLocalisedRecentChanges(configs.WhatsnewsDir)
+		if err != nil {
+			return fmt.Errorf("failed to read whatsnews, error: %s", err)
+		}
+
+		var releaseNotes []*androidpublisher.LocalizedText
+		for language, recentChanges := range recentChangesMap {
+			releaseNotes = append(releaseNotes, &androidpublisher.LocalizedText{
+				Language:        language,
+				Text:            recentChanges,
+				ForceSendFields: []string{},
+				NullFields:      []string{},
+			})
+		}
+		release.ReleaseNotes = releaseNotes
+		log.Infof("Update listing finished")
+	}
+	return nil
+}
+
+// readLocalisedRecentChanges reads the recent changes from the given path and returns them as a map.
+func readLocalisedRecentChanges(recentChangesDir string) (map[string]string, error) {
+	recentChangesMap := map[string]string{}
+
+	pattern := filepath.Join(recentChangesDir, "whatsnew-*-*")
+	recentChangesPaths, err := filepath.Glob(pattern)
+	if err != nil {
+		return map[string]string{}, err
+	}
+
+	pattern = `whatsnew-(?P<local>.*-.*)`
+	re := regexp.MustCompile(pattern)
+
+	for _, recentChangesPath := range recentChangesPaths {
+		matches := re.FindStringSubmatch(recentChangesPath)
+		if len(matches) == 2 {
+			language := matches[1]
+			content, err := fileutil.ReadStringFromFile(recentChangesPath)
+			if err != nil {
+				return map[string]string{}, err
+			}
+
+			recentChangesMap[language] = content
+		}
+	}
+	if len(recentChangesMap) > 0 {
+		log.Debugf("Found the following recent changes:")
+		for language, recentChanges := range recentChangesMap {
+			log.Debugf("%v\n", language)
+			log.Debugf("Content: %v", recentChanges)
+		}
+	} else {
+		log.Debugf("No recent changes found")
+	}
+
+	return recentChangesMap, nil
+}
+
+// GetRelease gets a release from a track based on it's status.
+func GetRelease(config config.Configs, releases *[]*androidpublisher.TrackRelease) *androidpublisher.TrackRelease {
+	status := GetReleaseStatusFromConfig(config)
+	for _, release := range *releases {
+		if status == release.Status {
+			return release
+		}
+	}
+	newRelease := androidpublisher.TrackRelease{
+		Status: status,
+	}
+	if config.UserFraction != 0 {
+		newRelease.UserFraction = config.UserFraction
+	}
+	*releases = append(*releases, &newRelease)
+	return &newRelease
+}
+
+// UpdateRelease creates and returns a new release object with the given version codes.
+func UpdateRelease(configs config.Configs, versionCodes googleapi.Int64s, releases *[]*androidpublisher.TrackRelease) error {
```

Comment:
It would be more clear to avoid passing configs around. Could parse it before passing it to this function; pass in a release and releaesatus

Reply:
Multiple values are used from config, methods would be less readable if they would have +3-4 args. I choose to pass config, as in the future we might implement draft and halted release statuses, which would require additional config params.